Add in new ocean omega component

vanroekel opened the pull request:
This PR adds in the new omega component.  I have confirmed it sets up cases and builds successfully

Commit:
Add ww3 component directory from cesm

Commit:
Consolidate ww3 cpl directory and delete .svn directories

 -

Commit:
Update ww3 buildlib to agree with consolidated cpl directory structure

Commit:
Change order of servers in config_inputdata.xml

Commit:
Bug fixes to get case.build to work with ww3

 - Fix syntax mistake in ww3 buildlib

 - Change name of user_nl_ww to user_nl_ww3

 - Update ww3 buildnml with correct component name and user
   namelist file name

Commit:
Add test model grid to config_grids.xml for wave+datm case

Commit:
Change parameters in namelist_definition_ww3.xml

 - Switch restart file so model starts from calm conditions

 - Set output frequency to 3 hours

Commit:
Add "only" option to ww3 module use statements in wav_comp_mct.F90

 - This makes it more clear which variables are needed in the
   coupling interface and helps inform what changes are necessary
   for incorporating the latest version of ww3

Commit:
Add cesm_splice_maps.py for creating ocean/wave remapping files

 - Takes in both a  bilinear and nearest neighbor remapping file
   and uses bilinear for the interior ocean and nearest neighbor
   in coastal regions.

   This avoids coupling issues between the waves and ocean caused by
   using bilinear interpolation at the coast.

 - Provided by Qing Li

Commit:
Modify ww3 NetCDF output in w3iogomd.f90

 - Make output consistent with newer version

Commit:
Remove unused subroutine from use statement in wav_comp_mct.F90

Commit:
Reduce output interval to 1 hour in ww3 namelist_definition_ww3.xml

Commit:
Switch output variables in wav_comp_mct.F90

 - Turn on peak frequency and direction

 - Turn off mean period and direction

Commit:
Add WW3 v5 source code

 - Only add files in the bin, ftn and inp directories for now

Commit:
Comment out coupling variables from CESM

 - Will add these back in later

Commit:
Add WW3 v5 compliation to buildlib

 - Pre-processes wavewatch *.ftn files

 - Also, complies wavewarch executables (not necessary, this step
   should be taken out later)

 - Calls E3SM Makefile to create libwav.a

 - Libaray creation fails right now because of version incompatibilites
   with exsiting mct interface (wav_comp_mct.F90)

Commit:
Update wav_comp_mct.F90 for WW3 v5

 - Comment out Langmuir enhancement (lamult) and mixed layer depth
   (HML) variables, since they are not in the new wavewatch source
   code.

 - Modify the field output specification. The outputs are now grouped
   into 9 different groups. The flag array to turn them on/off is
   now 2D

 - Add field output flag variables for coupling variables (not used)

 - Update call to w3init

 - Increase size of odat variable to account for coupling variables

Commit:
Use the correct mod_def.ww3 file in buildnml

Commit:
Add WW3 F90 source code files

1

Commit:
Get WW3 gridded output working

 - The NOTYPE variable wasn't being assigned and defaulted to 0.
   The loop that goes over output types to call the write routines
   was being skipped over.

 - Setting the NOTYPE variable to 7 in wav_comp_mct.F90 fixes the
   issue.

Commit:
Update WW3 bin directory after reinstall

Commit:
Update MPI complier name in WW3 comp and link files

Commit:
Don't remove *.f90 (lowercase f) files when running w3_clean

Commit:
Minor updates to WW3  buildlib

 - Add w3_clean command

 - Get rid of redundant environment variable specification

 - Switch NETCDF_CONFIG variable to intel

Commit:
Update pre-processed WW3 .F90 files for new switch configuration

Commit:
Move w3cesmmd.f90 to WW3 cpl directory

 - This is the cleanest way to include this file in the build
   process

Commit:
Revert "Don't remove *.f90 (lowercase f) files when running w3_clean"

This reverts commit 3a4455f081848e31f193b8adfd8e5c19396a34f1.

Commit:
Change WW3 link and comp files to gnu

Commit:
Update WW3 buildlib

 - Use WW3's w3_source script to pre-process the source code
   for the ww3_shel program only, instead of using w3_make to
   pre-process and compile all programs

 - w3_source creates a .tar.gz file that must be untared before
   calling the e3sm makefile

Commit:
Change WW3 restart file output interval to 5 days

Commit:
Update WW3 *.F90 files

 - Only include ww3_shel files since those are now the only ones
   generated by buildlib

 - ww3_grid files are also included, since that program was built
   using w3_make to create the mod_def.ww3 file

Commit:
Add the wwatch3.env file

Commit:
Update build scripts and coupler interface for WW3 v6.07

 - Add argument to call to w3init in wav_comp_mct.F90

 - Update paths to ww3 repo in buildlib. Also handle setting up the
   wwatch3.env file (if it doesn't already exist) by calling w3_setup

 - Point to correct mod_def.ww3 and restart.ww3 files in buildnml and
   namelist_definition_ww3.xml

Commit:
Comment WW3 buildlib and remove debug output

Commit:
Add switch_E3SM to control WW3 model options

 - This file will get copied to WW3 repo and when  w3_setup is run,
   it will cause this to be used as the switch file for pre-processing
   the .ftn source code

Commit:
Add some file removal cleanup to WW3 buildlib

 - This makes sure that the WW3 git working directory is clean (no
   untracked files) after building

Commit:
Remove ww3 v5 directory

Commit:
Move old CESM WW3 .f90 files to ww3_cesm directory

Commit:
Remove all .svn directories

Commit:
Updates to WW3 buildlib

 - Pass in extra variables when invoking make

Commit:
Fix compliation bug in wave_comp_mct.F90

 - Add additional variables in use statements

Commit:
Change wave model component name from ww to ww3

Commit:
Setup test wave compsets

Commit:
Add WW3 github repo as submodule

Commit:
Partially add wave radiation stress terms (Sxx, Sxy, Syy)

 - Add fields in seq_flds_mod.F90

 - Set w2x indicies in ww3_cpl_indicies.F90

 - Set x2o indicies in mpaso)cpl_indices.F90

 - Pack variables in wav_comp_mod.F90

 - No modifications to MPAS-O source code yet

Commit:
Add wave radiation stress terms to ocn_comp_mct.F

 - Unpack variables from attribute vectors

Commit:
Include ssh as a coupled variable between wav and ocn

Commit:
Changes to config_compilers.xml and config_machines.xml

Commit:
Fix typo in ocn_comp_mct.F

Commit:
Add compsets and grids for datm/wav testing

 - Use unstructured WaveWatchIII meshes

Commit:
Add CFSv2 wind config

Commit:
Update WW3 source to global unstrctured branch

Commit:
Update ww3 buildlib to include compling ww3_grid

Commit:
Update ww3 switch_E3SM

Commit:
Update buildnml for ww3

Commit:
Configure buildnml for ww3_grid

Commit:
Build additional namelist for ww3_grid

 - Also run ww3_grid in buildlib

sbrus89 committed:
Begin to add new WW3 compset

sbrus89 committed:
Updates to MPAS-O coupling fields

 - Remove radiaiton stress
 - Add stokes drift

sbrus89 committed:
Update WW3 coupling fields

 - remove radiation stress

sbrus89 committed:
Finish adding wQU100 and bug fixes

 - read from ww3_in instead of wav_in in wav_comp_mct.F90

 - remove UOST from switch file for now

sbrus89 committed:
Update WW3 submodule

sbrus89 committed:
Fix w3_setup input file in ww3 buildlib

Commit:
Move ww3_grid build after main make call in ww3 buildlib

Commit:
Change name of cime directory

Commit:
Add waves to cime_config files

Commit:
Remove cime_add-ww3

Commit:
Update description in ww3 config_component.xml

sbrus89 committed:
Bug fix for UWAV-SANDY compset long name

jgfouca committed:
Add cmake buildlib file, remove legacy buildlib

jgfouca committed:
Progress getting ww3 to build

jgfouca committed:
Comment to indicate where mesh.msh file is needed

jgfouca committed:
Add more comments

sbrus89 committed:
Comment out ww3_grid build

nairita87 committed:
Added compset F20TRC5-CMIP6-WW3 for Wavewatch with F case

sbrus89 committed:
Update WW3 submodule and related changes

 - E3SM WW3 branch has been rebased on most recent NOAA-EMC develop branch
   (fc3576f413acfe44f4c81e95eefd8ff7b5b21cdd)

   On top of this there are two important modifications:

     1) Fixes for rotated global unstructured meshes
     2) Re-organization of ww3_grid code into new module and subroutine

 - wav_comp_mct.F90 has been modified to update the interfaces to w3init
   and w3wave from the develop branch.

 - buildlib_cmake has be updated for the develop branch, w3_setup no
   longer uses the -c argument

 - buildlib_cmake now runs w3_source on the ww3_grid target to generate
   source code for the  w3gridmd module used to call w3grid from within
   wav_comp_init

 -  The wave_comp_init subroutine in wave_comp_mct.F was modified to
    call w3grid to generate the mod_def.ww3 and other related files
    prior to running the wave model. This is done serially.

sbrus89 committed:
Update compset for wQU225EC60to30 grid

sbrus89 committed:
Remove ww3_shel and ww3_grid .F90 files before building

sbrus89 committed:
Update build-namelist

 - Add path to .msh file for specific grid in ww3_grid.nml

 - Add path to station file

 - Turn on k propagation (for current coupling)

 - Reduce minimum depth to .25

 - Add ww3.input_data_list output

sbrus89 committed:
Add station output capability in wav_comp_mct.F90

sbrus89 committed:
Rotate winds into rotated wave coordinate system

sbrus89 committed:
Add namelist options for RTD grid

sbrus89 committed:
Update E3SM WW3 switches

sbrus89 committed:
Add CFSR_wQU225EC60to30 resolution

sbrus89 committed:
Add CFSR datm compset

sbrus89 committed:
Add wavewatch output and coupling flags to namelist

sbrus89 committed:
Rotate stokes drift vectors to standard pole

sbrus89 committed:
Clean up wav_comp_mct.F90 screen output (restrict to iaproc)

sbrus89 committed:
Adjust w2x variables for rotated grid

 - localize AnglD variable

 - rotate Stokes drift vector back to standard pole

 - switch to dedicated output processor to prevent overwriting
   local mean wave properties on ouptur processor

sbrus89 committed:
Add Stokes drift output to ww3_in namelist

sbrus89 committed:
Fix typo in stokes drift variable name in ocn_comp_mct.F

sbrus89 committed:
Add support for T62_oEC60to30v3_wQU225EC60to30 grid

 - Remove wDE4

sbrus89 committed:
Add hooks to enable active wave in mpas-o

sbrus89 committed:
Add pe layout for active wave component in current low-res G case

sbrus89 committed:
Clean up wav_comp_mct.F90: remove outfreq and histwr

sbrus89 committed:
Improve WW3 I/O

 - Make point and gridded output intervals namelist parameters

 - Write out restart files on driver interval

 - Read in restart files for continue and branch runs

sbrus89 committed:
Make ww3 restart equilvant to manual linking restart.ww3

sbrus89 committed:
Implement BFB restarts with ocean couplings

vanroekel commented:
@jonbob I'm a bit surprised by the huge # of commits.  Do you think this comes from the new cime tag?  It seems they are extra and unrelated as there are mmf/lnd/mpas PRs.  Should I squash them all or make a new branch and pick my commits on to that?

vanroekel commented:
Update -- I decided to squash commits to clean up the history.  But the files changed will be huge so I'd appreciate if one of you could check out the branch and try to build a case with OMEGA, I used GOMEGA-JRA, but feel free to choose another.

sbrus89 commented:
@vanroekel, sure. I can check it out and make sure cases build.

jonbob commented:
@vanroekel - I'm sure it's a mess between all the restructuring of CIME and MPAS and who knows what else... If it bothers you, you could try rebasing?

vanroekel commented:
Thanks @jonbob, I think you are right.   I think it is fine to just leave as is and pull in if @sbrus89 is able to build a couple cases.

Commit:
Update cime submodule hash and url/branch in .gitmodules

Commit:
Add ocean to wave ssh field in E3SM's seq_flds_mod.F90

Commit:
Add CFSv2 and CFSR datm configurations to E3SM datm

sbrus89 commented:
@vanroekel, I'm getting an error building `GOMEGA-JRA` with intel on Anvil. So far I'm not seeing anything very useful in the bldlog, but I'll keep digging a bit. What machine and compiler did you do your testing on?

sbrus89 commented:
Looks like this is the error:

*** No rule to make target `/gpfs/fs1/home/sbrus/E3SM/components/omega/src/core_seaice/column/constants/cesm/ice_constants_colpkg.F90', needed by `core_seaice/column/constants/cesm/ice_constants_colpkg.f90'

vanroekel commented:
Thanks for the testing @sbrus89 could you provide you create case line?  I'll try a clean checkout

vanroekel commented:
@sbrus89 I see the error too in a clean checkout.  I think I have a fix, testing now.

vanroekel committed:
Fixes sea ice build error in omega

vanroekel commented:
okay I think this should work now.  Please give it a try again.

sbrus89 commented:
@vanroekel, that fixed the issue. I'm able to build successfully now.

rljacob commented:
Is omega just an ocean component?  I notice omega/src has core-seaice, core-landice directories.

Will you eventually need a components/omega-ocean directory like the current components/mpas-ocean directory?

vanroekel commented:
@rljacob omega will eventually be ocean only, in essence replacing components/mpas-ocean but in a later iteration we will also make mpas-seaice a part of omega instead of its own component.

My thinking was to copy everything from mpas-source to allow tests to work easily at first and then slowly factor out the ocean code as things get replaced.  I need the whole directory as we plan to completely rebuild mpas-source/src/framework which would break MPAS builds. I'm talking with Jim F. about how best to factor the code in a follow on PR to this addition.

jonbob committed:
Subtree merge from MPAS-Model hash in OceanNGD/add-ww3

jonbob committed:
Update bld script to match Registry changes

jonbob committed:
Merge branch 'e3sm/sbrus/ocean/support-ww3' into jonbob/ocean/rebase-add-ww3

* e3sm/sbrus/ocean/support-ww3:
  Subtree merge from MPAS-Model hash in OceanNGD/add-ww3

jonbob committed:
Update CIME version to a branch for supporting ww3

jonbob committed:
Clean-up from subtree merge

jonbob committed:
Re-merge branch 'e3sm/sbrus/ocean/support-ww3' into jonbob/ocean/rebase-add-ww3

sbrus89 committed:
Update cime branch to jonbob/e3sm/support-add-ww3 in .gitmodules

vanroekel committed:
Fixes the WCYCL1850_OMEGA compset

vanroekel commented:
@sbrus89 I believe I've fixed the `WCYCL_1850` case, could you test again?  It worked for me on anvil.  I'll work on rebasing this to master next so we can hopefully merge this soon.

vanroekel committed:
updates cime

vanroekel committed:
Adds omega ocean component

Conflicts:
	.gitmodules
	cime
	cime_config/allactive/config_pesall.xml
	components/cmake/build_model.cmake
	components/mpas-source

vanroekel committed:
updates cime and pe layouts

vanroekel commented:
@sbrus89 I finally got this rebased.  I tested a b-case with omega and it was okay.  This PR now needs to wait for your ww3 PR to go in.

sbrus89 commented:
@vanroekel, that great! I'll get the ww3 PR submitted asap.

vanroekel commented:
shoot, I'm getting build failures in elm now related to missing links to some cime stuff.  I assume this is a rebase issue.  I'm going to try tart clean and see if I can get past it.

vanroekel committed:
updates cime and fates to fix build errors

vanroekel commented:
okay, I've worked past the build errors.  @sbrus89 could you take a test drive when you have a chance?

sbrus89 commented:
@vanroekel - yes, will do.
